### PR TITLE
Add HLL boundary correction to GRMHD

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

Add the HLL boundary correction for the GRMHD system

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryCorrections/Hll.hpp [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+ * where \f$\lambda^{+}\f$ (\f$\lambda^{-}\f$) is the largest characteristic
+ * speed in the outgoing (ingoing) direction. Note the minus signs in front of
+ * \f$\lambda^{\pm}_\text{ext}\f$, which is because an outgoing speed w.r.t. the
+ * neighboring element is an ingoing speed w.r.t. the local element, and vice
+ * versa. Similarly, the \f$F_{\text{ext}}\f$ term in \f$G_\text{HLL}\f$ has a
+ * positive sign because the outward directed normal of the neighboring element
+ * has the opposite sign, i.e. \f$n_i^{\text{ext}}=-n_i^{\text{int}}\f$.
+ *
+ * The characteristic/signal speeds are given in the documentation for
+ * grmhd::ValenciaDivClean::characteristic_speeds(). Since the fluid is
```
Comment (moxcodes) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryCorrections/Hll.hpp]: [nitpick] backticks for the function

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryCorrections/Hll.hpp [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+ * travelling slower than the speed of light, the speeds we are interested in
+ * are
+ *
+ * \f{align*}{
+ *   \lambda_{\pm}&=\pm\alpha-\beta^i n_i,
```
Comment (moxcodes) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryCorrections/Hll.hpp]: In this equation the `\pm` is in the subscript, where in the previous descriptions they're in the superscript. I think they mean the same thing, so maybe just make this one a superscript. If it denotes something different, then the prose should clarify.

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryCorrections/Hll.cpp [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+  normal_dot_flux(packaged_normal_dot_flux_tilde_s, normal_covector,
+                  flux_tilde_s);
+  normal_dot_flux(packaged_normal_dot_flux_tilde_b, normal_covector,
+                  flux_tilde_b);
+  normal_dot_flux(packaged_normal_dot_flux_tilde_phi, normal_covector,
+                  flux_tilde_phi);
+
+  using std::max;
+  return max(max(abs(get(*packaged_largest_outgoing_char_speed))),
+             max(abs(get(*packaged_largest_outgoing_char_speed))));
```
Comment (moxcodes) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryCorrections/Hll.cpp]: one of these was probably supposed to be ingoing, I think.

**Comment by nilsdeppe** [comment]
Rebased and squashed in changes since they're relatively small. Thanks for the review :)

**Review by moxcodes — DISMISSED** [review]

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryCorrections/Hll.hpp [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+class not_null;
+}  // namespace gsl
+namespace PUP {
+class er;
+}  // namespace PUP
+/// \endcond
+
+namespace grmhd::ValenciaDivClean::BoundaryCorrections {
+/*!
+ * \brief A Rusanov/local Lax-Friedrichs Riemann solver
```
Comment (kidder) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryCorrections/Hll.hpp]: RIP

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryCorrections/Hll.cpp [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+#include <optional>
+#include <pup.h>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tags/TempTensor.hpp"
+#include "DataStructures/Tensor/EagerMath/DotProduct.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "NumericalAlgorithms/DiscontinuousGalerkin/Formulation.hpp"
+#include "NumericalAlgorithms/DiscontinuousGalerkin/NormalDotFlux.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
```
Comment (kidder) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryCorrections/Hll.cpp]: unused

File: tests/Unit/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryCorrections/Test_Hll.cpp [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+#include "Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryCorrections/Factory.hpp"
+#include "Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryCorrections/Hll.hpp"
+#include "Evolution/Systems/GrMhd/ValenciaDivClean/System.hpp"
+#include "Framework/SetupLocalPythonEnvironment.hpp"
+#include "Framework/TestCreation.hpp"
+#include "Helpers/Evolution/DiscontinuousGalerkin/BoundaryCorrections.hpp"
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "PointwiseFunctions/Hydro/EquationsOfState/IdealFluid.hpp"
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryCorrections/Test_Hll.cpp]: unused

File: tests/Unit/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryCorrections/Test_Hll.cpp [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+#include "Utilities/TMPL.hpp"
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryCorrections/Test_Hll.cpp]: unused

**Review by kidder — CHANGES_REQUESTED** [review]
you can squash

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryCorrections/BoundaryCorrection.hpp [hunk 8]
```diff
@@ -12,6 +12,7 @@
 /// Boundary corrections/numerical fluxes
 namespace grmhd::ValenciaDivClean::BoundaryCorrections {
```
Comment (kidder) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryCorrections/BoundaryCorrection.hpp]: oh and this is doxygen version dependent on whether `BoundaryCorrections` is documented

**Review by kidder — CHANGES_REQUESTED** [review]

**Comment by nilsdeppe** [comment]
Rebased, squashed, and added a commit moving the namespace comment. Thanks for the reviews :)

**Review by kidder — APPROVED** [review]

**Review by moxcodes — APPROVED** [review]

**Pull request merged**